src/libopensc/card-starcos.c

```diff
@@ -2160,18 +2160,9 @@ static int starcos_card_ctl(sc_card_t *card, unsigned long cmd, void *ptr)
 	}
 }
 
-/**
- * starcos_logout_v3_x()
- * StarCOS 3.x cards will not clear the security status by selecting MF.
- * Returning NOT_SUPPORTED would cause card reset, effectively invalidating
- * the security status.
- */
```

Comment:
This comment looks like it was never true as returning `NOT_SUPPORTED` is mostly ignored by the callers of the logout function. OTOH, would it make sense to make it this way to avoid calling the `sc_reset()` in most of the readers and just return `NOT_SUPPORTED` and let the something outside of the driver handle the situation (call the `sc_reset()`)?

Reply:
Indeed, a reset was never done if logout was *not* implemented, so the comment is and has always been wrong. And I think it would be a bad idea to implement it this way.

I agree that propagating the error would be the most sane thing to do as it may be handled in some situations (see comment above about handling it in CTK).

Reply:
That being said, the purpose of this PR is to actually *do a reset* if we don't know better what to do and if we know it is safe to reset.

Reply:
If you don't know what to do, look at `disconnect_action`

Reply:
> If you don't know what to do, look at `disconnect_action`

I do not think the disconnect action should be executed on logout without any warning (even as a fallback).

Reply:
What I was implying is if `opensc.conf` has `disconnect_action = reset;` or `disconnect_action = unpower;` then user has modified  the parameter from default of `disconnect_action = leave;` in response to:
> That being said, the purpose of this PR is to actually do a reset if we don't know better what to do and if we know it is safe to reset.

Note that pkcs11 3.0 says: "C_Logout logs a user out from a token."  "Depending on the current user type, if the call succeeds, each of the application’s sessions will enter either the “R/W Public Session” state or the “R/O Public Session” state."  The login state is then controlled by the PKCS11 middleware, just as when   

I read this as this application still has access to the token with "Public Sessions" so reset should not be used. And other applications may still continue to have access to the token wit a logged in state.

PKCS11, PCSC and ISO 7816 don't match when multiple applications have their own PKCS11 and are using shared access to a token via PC/SC. 

But OpenSC does define the  `disconnect_action` that imply (leave) if login state should be not be changed or (reset) which would cause the token to lose the login state.    There could be a 4th state `disconnect_action = logout;`   which would then call the iso7816 logout if user wanted to change the token login state.

Reply:
Let me express the strongest agreement with the point @dengert made.

Reply:
`disconnect_action` does not yet have this security related connotation, I doubt that using this for logout would be an easy to get change for most users.

I understand that we may return an error to the application when a logout is requested (currently, we actually don't). But I don't think we need to put this burden onto the application. (Also note that an application cannot make  a reset via PKCS#11) My thought process e.g. for PIV is the following:
1. If we know how to manually log out via an APDU, go ahead and do so
2. if that fails and it's an applet based card, reselect the applet, which *typically* resets the security state (essentially I'm performing the action from the card's `match_card` and `init` is performed).
3. if that fails or cannot be applied, use `sc_reset` as last resort, which, in particular, is safe to use on static card that don't have volatile data loaded at `init` and `match_card`

I don't think that this would cause problems for applet based cards that have multiple opened sessions to different applications. For one, I cannot imagine a use case scenario where OpenSC is used somewhere and we don't have any information about anyone actually using our middleware this way. Second, assume some middleware supports concurrent access *and* is used together with OpenSC, then we are still safe to use a reset, because that other middleware would need to have some support for recovering from a reset (what we have implemented in `card_reader_lock_obtained`). If it doesn't have such thing implemented, then that middleware likely cannot really be used in parallel with OpenSC or anything else.

In conclusion, we are choosing a reset only if we don't have a any other choice and it should be safe to use. If you see any problems arising from this, please elaborate on the use case and on the software/hardware involved in this.

Reply:
> disconnect_action does not yet have this security related connotation, I doubt that using this for logout would be an easy to get change for most users.

Well then add something like `logout_action = yes|NO`  Give the users a choice to continue using the cards as they have done since 2004.  See https://github.com/OpenSC/OpenSC/pull/2807#discussion_r1240997601 Don't break what has been working. Give users a choice even if it means another `opensc.conf` option. 

My concern is you will be turning over an application the choice to reset login state or not. Applications like FireFox, Thunderbird or others. Any one could reset the login state at any time.    

> My thought process e.g. for PIV is the following:

Why PIV only?

>If we know how to manually log out via an APDU, go ahead and do so 

NIST did not introduce logout until 800-73-4 in 2016 and the use of "leave" has worked well to not require pin caching and allow the use of PIN PAD readers where a pin can't be cached.     

>if that fails and it's an applet based card, reselect the applet, which typically resets the security state (essentially I'm performing the action from the card's match_card and init is performed).

PIV are always an "applet based card" and NIST has always said reselecting the PIV applet does not reset the login state. (But some "PIV Like" cards have not followed the standard.) Only if some other applet that exists on the card is selected, then the PIV login state is reset. So you would have to know what applets are on the card to use this method. 

>if that fails or cannot be applied, use sc_reset as last resort, which, in particular, is safe to use on static card that don't have volatile data loaded at init and match_card 

Like `disconnect_action = reset;` vs  `disconnect_action = unpower;`  Which cards does OpenSC support that are not static? 

In regard to cards with multiple applets used concurrent: We have that now with OpenPGP and PIV cards, or CAC and PIV cards. Its just the OpenSC can only support one of the applets from the same process, but users such as @mouse07410  I believe used other software for one of the applets. Starting an application with OPENSC_DRIVER= can also cause OpenSC in different processes to different applets.  So this is a case for pin caching by each long running application.

Reply:
AFAIK, nobody is using multiple card applications *in parallel*. Multiple long running processes accessing two different card applets at the same time are technically possible, but that's just an illusion. Organizations and people are fully occupied handling crypto tokens with just a single key, they don't want it to be even more complex than that. At least, that's my experience. If you, however, find some use case that actually breaks by accepting this pull request, then please let me know and we can think of some more complex solution.

CTK, MD and PKCS#11 created complex infrastructure to separate processes from each other that are accessing a single card at the same time to avoid a misuse of the card and you want to defeat all this by just leaving the card authenticated even after explicit logout? Look, here is the super secret key to your house, but anyone pressing the door handle can enter as well. I would consider this a security problem!

What NIST thought in their initial specification? Maybe they just forgot, maybe they thought a reset is good enough. I don't know and I don't care. I used PIV just as an example, the security issue I'm talking about, touches **every single one of those 22 different types of tokens from this pull request**.

Reply:
@frankmorgner  I have given you my arguments, and options to give uses/administrators a choice. Do wat you want.

Reply:
OK, understood.

@Jakuje , what's your opinion on using `sc_reset()` for logout for some of the cards and/or adding a new configuration option for this (and its default value)?

Reply:
Using multiple applets on a card/token is broken and will likely be broken forever. Unfortunately, there is yubikey which presents openpgp and piv applet. The GnuPG using OpenPGP and the rest of the world is using OpenSC with PIV default will be breaking always until either GnuPG will start accessing the card through PKCS#11 or I dont know what.

It is correct that in this case or in case several application will access the single applet on a card/token the reset will forcibly logout any other application, which is something the applications might not be expecting and might not handle correctly and which could cause quite huge influx of issues and we need some fallback for. Hardcoding `sc_reset()` into the card drivers will not make this easier for us.

On the other hand from my experience, the `C_Logout` is not called very frequently by applications. They usually call `C_Login` and then keep the login state until they call `C_Finalize` so I believe most of the users and applications will really not notice this change as their application will got through `disconnect_action` only.

My proposal would be to go ahead with returning `SC_ERROR_NOT_SUPPORTED` instead of calling the `sc_reset()` from the card drivers and handle this in one place.

How, it is another question. We need sane/secure defaults and we need "compatible" option which sounds like the only way to achieve this is with another opensc.conf option, `logout_fallback = [reset|leave|...]` that would allow use to configure this behavior separately from the disconnection action.

Reply:
OK, I've removed all manual calls to `sc_reset()`.

For later enhancement for logging out, please note the following:
- In Minidriver, Windows will perform a reset on its own if [CardDeauthenticateEx](https://learn.microsoft.com/en-us/previous-versions/dn468713(v=vs.85)) (`sc_logout()`) fails.
- In PKCS#11 we're calling `sc_logout()` also when a session terminates.